clang/lib/Sema/SemaExpr.cpp

```diff
@@ -13736,6 +13756,42 @@ QualType Sema::CheckAssignmentOperands(Expr *LHSExpr, ExprResult &RHS,
     ConvTy = CheckAssignmentConstraints(Loc, LHSType, RHSType);
   }
 
+  // __builtin_counted_by_ref can't be used in a binary expression or array
+  // subscript on the LHS.
+  int DiagOption = -1;
+  auto FindInvalidUseOfBoundsSafetyCounter = [&](Expr *E) {
+    struct BuiltinCountedByRefVisitor
+        : public RecursiveASTVisitor<BuiltinCountedByRefVisitor> {
+      MemberExpr *ME = nullptr;
+      bool InvalidUse = false;
+      int Option = -1;
+
+      bool VisitMemberExpr(MemberExpr *E) {
+        if (auto *FD = dyn_cast<FieldDecl>(E->getMemberDecl());
+            FD && FD->isBoundsSafetyCounter())
+          ME = E;
+        return true;
+      }
+
+      bool VisitArraySubscriptExpr(ArraySubscriptExpr *E) {
+        InvalidUse = true;
+        Option = 0; // report 'array expression' in diagnostic.
+        return VisitStmt(E->getBase()) && VisitStmt(E->getIdx());
```

Comment:
I believe you don't need `VisitStmt` here because `RecursiveASTVisitor` will do it for the subexpressions as long as this visitor returns true.